### PR TITLE
Cleaned up Chunk.relightBlock

### PR DESCRIPTION
**Pull request by Mathe172**
This PR is the second one in our attempt to fix and optimize the vanilla lighting engine (first one: #3870). For more details see #3848. As suggested there, we'll try to break the changes into as many PRs as possible.

This particular PR speeds up `Chunk.relightBlock` by removing some unnecessary stuff and fixes part of the issue described in [MC-46129](https://bugs.mojang.com/browse/MC-46129) (see the second image: placing a platform over a forest causes the leaves of the trees to stay bright; another part of the issue is fixed by #3870) **[Edit]** See [MC-117067](https://bugs.mojang.com/browse/MC-117067) for an issue describing only this bug **[/Edit]**

| | What vanilla does | What we do | Remarks/Description |
| --- | ---- | ---- | ---- |
| 1. | Recalculate the height of the column |  |  |
| 2. | Call to `World` `.markBlocksDirtyVertical` | Remove it | This marks the block for rendering (which is already done for light updates) and recalculates the light for the affected part of the column (which is useless, as the `heightMap` is only updated after the call)|
| 3. | | Call the new `LightingHooks` `.relightSkylightColumn` | See below |
| 4. | Try to guess the light values of the columns | Bypass it | This essentially breaks the checks that come afterwards, which is part of the cause of [MC-117067](https://bugs.mojang.com/browse/MC-117067) |
| 5. | Schedule checks for column + 4 neighbors | Bypass it | It suffices to recalculate the column, provided the light map is not messed up by the above. Since the light values are wrong, the fact that the block below is not checked causes [MC-117067](https://bugs.mojang.com/browse/MC-117067) |

The lighting engine correctly calculates the source light values for the sky light (i.e. 15 above the first non-transparent block, and 0 below) - this means that nothing needs to be calculated manually.

What `LightingHooks.relightSkylightColumn` does:
* Just let the affected part of the column be rechecked by the lighting engine, which handles everything correctly
* For empty sections, we need to look at the neighboring blocks too, as there are no old light values to be checked against (since `chunk.getLightFor` simply looks at the `heightMap` for empty sections, light changes won't be detected)

Regarding [line 97](https://github.com/MinecraftForge/MinecraftForge/compare/1.11.x...Mathe172:lighting_cleanup?expand=1#diff-496aa22ea49bec02cab69d859142959dR97):
`Chunk.propagateSkylightOcclusion` indirectly calls `recheckGaps`, which is supposed to (as far as we can judge) fix issues in the light map. Those can be caused by one of the following:
* Neighboring chunks are not loaded. `recheckGaps` can't do anything in this case - while it would wait for the neighboring chunks to be loaded, it forgets this as soon as the chunk itself is unloaded, which makes the "fix" very unreliable in the first place. During world-gen, this call is irrelevant, as the checks are scheduled by `Chunk.checkLight`. Also, the condition above the call is hard to achieve and doesn't apply in many of the cases where it should.
* `Chunk.generateSkylightMap` is called - since that is broken anyway (and fixed in #3870) and isn't fully fixed by `recheckGaps`, this is also irrelevant
* `Chunk.relightBlock` is called, which is fixed in this PR.

Remarks:
* This PR removes 5/6 of the checks of `Chunk.relightBlock` and all of `recheckGaps`, while also making it more understandable. (We know the checks are very cheap, but they take a big part of the time of the lighting engine)
* We created LightingHooks as we plan to add a few more fixes in the near future (see e.g. #3870). If needed, we can move it to somewhere else until there's enough for a new file.
* In order to see the correct result for [MC-117067](https://bugs.mojang.com/browse/MC-117067), you need to relog (because of some client syncing bugs, which we plan to address later)
* If you spawn a platform over a forest, there will be bright spots on the ceiling (and in the column below) (fixed by #3870)
* The commits are incompatible with #3870, as we didn't want them to depend on each other. We will regen them should one be accepted.

### CODE REVIEW

**Comment by williewillus** [comment]
Can we clearly document all that the vanilla code does + the changes somewhere (that's not here) so each new version of mc we can cross-check if something changes?

Cleaning up/optimizing the lighting is great but carries the big risk that we aren't up to date with vanilla changes if they happen / don't know if our reimplementations are correct

**Comment by bs2609** [comment]
Probably also nice to publish some of your findings on the official bug tracker - any vanilla fixes means less for Forge to maintain.

**Comment by Mathe172** [comment]
@bs2609 Good idea, [done](https://bugs.mojang.com/browse/MC-55730?focusedCommentId=381619&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-381619)

We also created a [new issue (MC-117067)](https://bugs.mojang.com/browse/MC-117067) for this bug, as none of the other ones identifies this as being different issue.

**Comment by williewillus** [comment]
I think the patches themselves that bypass vanilla calls should have a brief explanation why it's patched out so in the future people don't have to come dig up this PR to read about the change

**Comment by PhiProven** [comment]
Following @williewillus's suggestion, we added some comments to the patched lines.

**Comment by Mathe172** [comment]
@mezz Regarding your comment on #3888: Is there anything we could do to make testing/reviewing easier?

**Comment by mezz** [comment]
I should be more clear that there is nothing wrong with your PR, I think it's cool that you decided to dig into this. The issue is that it touches extremely suspect vanilla code, and the benefits have to outweigh the potential risks.

There may be trade offs or side effects of fixing these bugs that are difficult to find, because the vanilla lighting system appears to be fundamentally buggy and has been patched over [many times](https://bugs.mojang.com/issues/?jql=project%20%3D%20MC%20AND%20labels%20%3D%20lighting). Despite that, the existing system seems to have "ok" performance and the bugs are not considered game-breaking. To me this means the state of the code is very precarious. Ideally it might be rewritten by Mojang, but since it is good enough for now it is probably not a priority.

For inclusion in Forge the changes have to be shown to be safe and worth the effort, so unfortunately the bar is set very high here. The last thing we want is new lighting or performance bugs that are experienced more often than the uncommon ones fixed in this PR.
It would be helpful if you could look into potential bug regressions, and also investigate the possibility of performance differences in areas outside of the bugs you targeted.

File: src/main/java/net/minecraftforge/common/lighting/LightingHooks.java [hunk 1]
```diff
@@ -38,9 +38,12 @@ public static void relightSkylightColumn(final World world, final Chunk chunk, f
         final int xBase = (chunk.xPosition << 4) + x;
         final int zBase = (chunk.zPosition << 4) + z;
 
-        for (int y = yMax; y >= yMin; --y)
+        int y;
+        ExtendedBlockStorage section = null;
```
Comment (Niko-sk2x) [inline src/main/java/net/minecraftforge/common/lighting/LightingHooks.java]: why use `null `here and `Chunk.NULL_BLOCK_STORAGE` later when checking for null? (also, stupid github now shows me as "reviewer" even when all I wanted to do was point out this)

Comment (Mathe172) [reply src/main/java/net/minecraftforge/common/lighting/LightingHooks.java]: The check needs to be against `Chunk.NULL_BLOCK_STORAGE` - the initialization is only to keep the compiler from complaining (the value never matters) and `null` is just shorter to write ;-)

**Comment by Mathe172** [comment]
First of all, thanks for your answer.

Regarding your concerns:
In short: We took great care to ensure that the PR does not introduce any regressions compared to vanilla.
Long version:
Note that we didn't just fix the lighting bugs one after another - we looked at the lighting code, deduced what could go wrong (only after this we looked for corresponding vanilla bug reports) and finally fixed/cleaned up the code. (The code should now be logically clean - would be nice if we could get Mojang to implement this type of fixes directly instead of having to carry them from version to version ;-) )
The following remarks concern #3870, #3871 & #3879 (while they are independent of each other, it's easier to discuss them together to give a cleaner overview). We excluded both client PRs (#3891 & #3898) as they're not really related to the lighting engine and should be treated separately.

Regarding testing: It's not really possible to implement any unit tests for this kind of stuff, as these would never be able to check the edge cases that produce the issues.
As a theoretical "proof" why these PRs do in fact produce the expected results:
First, a quick overview over what the Vanilla lighting engine does:
* The core method is `World.checkLightFor`. This method detects and spreads the light changes. It does so by using a variant of a breadth-first search for spreading light (cached values are taken from `Chunk.getLightFor`). If a block gets darker, it finds all blocks potentially illuminated by it (using the same method) and spreads the light in from the borders of this region.
  * Our lighting engine (#3879) effectively does the same, it just delays those checks a bit and processes multiple of them simultaneously, the logic itself stays the same however.
* Every time some property of a block concerning lighting changes, `World.checkLightFor` needs to be called. For block light, this is trivially done inside `World.setBlockState`. For skylight, the situation is a bit more difficult:
  * Light sources are determined by looking at `Chunk.heightMap`: Everything below any block that's not fully transparent is dark (source value `0`), everything above has a source value of `EnumSkyBlock.SKY.defaultLightValue` (`15`).
  * This means that for every block change, the height of the respective column has to be determined and everything between the old and new height needs to be rechecked. This is exactly what `Chunk.relightBlock` does after this PR.
    * In order for `World.checkLightFor` to correctly detect light changes, `Chunk.getLightFor` needs to properly return the **old** light value. There are few places where this is not fulfilled:
      * `Chunk.generateSkylightMap`: It tries to guess (and directly sets) the light values by tracking the light downwards (which is wrong, as the light could spread in from the sides). This issue is resolved by #3870 (only for cases after world-gen)
      * Empty sections (i.e. `Chunk.NULL_BLOCK_STORAGE`): For these, the light value is calculated on-the-fly by looking at the height map, instead of relying on cached values. This means that light changes may not be detected in this case. Therefore, the neighbor blocks need to be check in addition to the block itself.
      * The vanilla version of `Chunk.relightBlock`: This method also guesses the light values manually instead of relying on `World.checkLightFor` and sets them directly. While this produces the correct result (as all neighbor columns are checked) (excluding the one block below the column that is missed), it is unnecessarily slow and unreadable (which is what this PR is for).
* This leaves room for two possible bugs:
  * If not enough chunks are loaded around a light change, the changes can not be propagated properly. Vanilla just does not process this kind of changes at all, leaving a single incorrect light value. We took a different route and spread the changes as far as possible (i.e. to the border of the loaded area). This seems more natural as it produces the correct result in case the changes do not need to be propagated that far and the defects are also easier to fix, as they can all be found on chunk boundaries (see #4184)
  * Light sources that are placed during world-gen (not population): For those, no checks are done, in particular no light checks. For population, the checks are done, effectively reducing this issue to the first point.
      * `Chunk.checkLight` attempts to fix (but does a terrible job) both these issues after world-gen + population and fixes issues from `Chunk.generateSkylightMap` (which is called once during world-gen)
      * Side note: Placing any complicated block (as light sources) during the initial generation is arguably a bad idea, as no checks are done. This means that the checks in `Chunk.checkLight` are more or less unnecessary and could be removed (as they take up a third of the world-gen time), but this is another discussion.

One small regression introduced by our version: The vanilla engine is better at fixing defects on-the-fly, as it doesn't rely on the values being correct as much as we do. This is arguably a bad and unreliable way to prevent bugs and should not be an issue if the bugs do not occur in the first place.

The most dangerous part of our PRs is probably the fact that `LightingEngine.procLightUpdates` needs to be called before light values are accessed.
* `Chunk.getLightFor`/`.getLightForSubtracted` cover most cases
* Accessing the underlying data arrays directly should only be done by saving and `SPacketChunkData`, which is unlikely to change in the future.
* The most critical part is `IWorldEventListener` as there is no general solution for that. For vanilla, only `RenderGlobal` has a non-trivial implementation, which we took care of. (Should we add a comment to `IWorldEventListener.notifyLightSet` to emphasize that light changes should be processed beforehand?)

**Comment by Mathe172** [comment]
Another thing we should mention/clarify:
There is one possible regression of our implementation of the lighting engine compared to vanilla: Since the lighting is spread to the borders of the loaded region during world-gen, `Chunk.checkLight()` has no way to detect the defects (as the actual light source itself is correct). This issue is corrected in #4184 (see previous comment).
For now, there are two options:
* Leave it like it is - while there could theoretically be a regression, we were unable to see any such defects in our testing, only improvements (as `Chunk.checkLight` itself does a terrible job, especially in the nether, see for example [MC-3329](https://bugs.mojang.com/browse/MC-3329))
* Don't do the spreading to the boundaries for now and change this when we make the boundary check PR, where this spreading is actually needed and will never cause regressions

**Comment by PhiProven** [comment]
@mezz Any updates? The PRs can really be treated separately
* #3870 is just a small bugfix. This should be the simplest one
* This one (#3871) is a cleanup and thus requires a basic understanding of how Vanilla lighting works (which we have outlined in the PR description). Its main purpose is to simplify the code to make reviewing the more complicated PR easier, but it also fixes a bug and slightly improves performance.
* #3879 is the most complicated, but also has the biggest impact on performance. There are two separate parts to verify:
  * `LightingEngine` works as intended. This is the "easy" part, as there aren't any complicated interactions with minecraft. For this part one *could* write a unit test, but it's probably less work to just theoretically verify that the algorithm works.
  * Making sure that light updates are processed sufficiently often. This is probably the hardest part to check. We have listed the places where they need to be processed above.
* #4184 is also just a small bugfix. The idea behind it is quite simple. Unfortunately, the code got a bit long due to vector calculations.

**Comment by mezz** [comment]
As discussed with PhiPro95 on Discord, I am closing these lighting PRs in favor of PhiPro95 and Mathe172 implementing them in a separate mod that is dedicated to fixing lighting.
